### ISSUE TITLE
Ha Doan: Welcome to JS, 3 weeks 

### ISSUE THREAD

**Issue by hadoan149** [lab-brussels-1/home#318]
## Learning Objectives

<summary>Priorities: 🥚, 🐣, 🐥, 🐔 (click to learn more)</summary>
<br>

There is a lot to learn in this repository. If you can't master all the material
at once, that's expected! Anything you don't master now will always be waiting
for you to review when you need it. These 4 emoji's will help you prioritize
your study time and to measure your progress:

- 🥚: Understanding this material is required, it covers the base skills you'll
  need for this module and the next. You do not need to finish all of them but
  should feel comfortable that you could do them with enough time.
- 🐣: You have started all of these exercises and feel you could complete them
  all if you just had more time. It may not be easy for you but with effort you
  can make it through.
- 🐥: You have studied the examples and started some exercises if you had time.
  You should have a big-picture understanding of these concepts/skills, but may
  not be confident completing the exercises.
- 🐔: These concepts or skills are not necessary but are related to this module.
  If you are finished with 🥚, 🐣 and 🐥 you can use the 🐔 exercises to push
  yourself without getting distracted from the module's main objectives.

---

</details>

### [1. What is Programming](./1-what-is-programming)

What is a program? What is a programming language? How do programs and people
fit together?

- [x] 🥚 **Data in, Data Out**: You understand that computer programs are used
      to transform data, and can explain a JS program from this module in terms
      of the data that goes in and the data that comes out.
- 🥚 **Three Audiences**: You can explain how a single file of code is used to
  communicate with 3 different audiences:
  - [x] **Developers**: You can explain how code formatting, comments, logs and
        variable names make it easier (or harder!) for a developer to understand
        a program.
  - [x] **Computers**: You can explain how a computer follows your instructions
        to store, read and modify data in program memory.
  - [x] **Users**: You can explain how the computer creates a dynamic user
        experience by following different executions paths depending on user
        input.
- [x] 🥚 **Tracing Variables**: You can complete a trace table for a program
      that uses only variables, and can check your work using the "trace"
      button.
- 🥚 **JavaScript Syntax**: Given a program, you can identify these parts of
  JavaScript syntax:
  - [x] _Primitives_
  - [x] _Operators_
  - [x] _Identifiers_
  - [x] _Keywords_
  - [x] _Checks_
  - [x] _Blocks_
  - [x] _Function Calls_
- [x] 🥚 **Pseudo Code**: You can explain what pseudo code is, how it's used and
      why it's helpful.
- [x] 🥚 **Program Goals**: You understand that a full JavaScript program can be
      explained as a series of smaller goals, and can write names for the goals
      in a simple working program.
- 🐣 **Program Analysis**: You can study a simple working program and identify:
  - [x] The program's behavior (_data in, data out_)
  - [x] The program's goals
  - [x] The program's syntax and language features
  - [x] Which parts of the source code are for which audience (users, devs,
        computers)
- [x] 🐥 **Same Behavior, Different Code**: You can study simple programs with
      the same behavior (_data in, data out_) but different source code and
      explain how the differences matter for each audience.

### [2. Just Enough JavaScript](./2-just-enough-javascript)

Go in depth on JavaScript you need to know for writing interactive text-based
programs in the browser. Along the way you will learn how each language feature
works in small programs.

- 🥚 **Listening and Reading**: You can read code out loud, and understand your
  classmates when they read code to you. You don't need to understand how a
  program works to master these learning objectives!
  - [ ] **Listening** You can exactly re-write a program that a classmate has
        read
  - [ ] **Reading** You can read a program out loud and guide your classmates to
        re-write _exactly_ the same code without them seeing the program. Every
        indentation, semi-colon, comment and spelling in their program must be
        identical to yours.
- 🥚 **Static vs. Dynamic Analysis**: You can explain and use these two ways of
  studying a program, each can help you understand different aspects of your
  code. To help understand this concept, the _options_ panel in Study Lenses is
  organized into _static_ and _dynamic_ study options:
  - [ ] **Static**: Studying the text in a code file _without_ running it. Some
        static study methods are creating a flowchart, analyzing variables,
        filling out a trace table, and drawing on code.
  - [ ] **Dynamic**: _Running_ code and studying the computer's behavior. Some
        dynamic study methods are running code and reading console logs, using
        the _trace_ button, and stepping through in the debugger or JS Tutor.
- [ ] 🥚 **Tracing Execution**: You can complete a "steps" trace table for all
      JS language features in Just Enough JavaScript, and can correct your table
      using console output from the "trace" button.
- [ ] 🥚 **Analyzing Variables**: You can list all the variables in a program,
      and answer these 5 questions for each variable:
  - _Where is the variable declared?_
  - _What is the variable's scope?_
  - _Is the variable initialized with a value?_
  - _How many times is its value used \(read\) in the program?_
  - _How many times is the variable assigned a new value?_
  - _What types are assigned to this variable during the program's execution?_
- [ ] 🥚 **Completing Programs**: You can successfully fill in blanks for a
      program when the missing words are provided, including distractors.
- [ ] 🐣 **Translating Pseudo Code**: Given a program written in Pseudo Code,
      you can translate it to working JavaScript.
- [ ] 🐣 **Comparing Programs**: You can compare two programs with similar code
      and explain if they have the same behavior or not. If they do not, you can
      explain how they behave differently and why.
- [ ] 🐥 **Constructing Programs**: You can reconstruct a program's lines and
      indentation, successfully ignoring distractor lines.
- [ ] 🐥 **Linting**: You can find and fix simple linting errors in JavaScript
      programs.
- [ ] 🐔 **Writing Programs**: You can write small programs that log specific
      values, or pass specific test cases.

### [3. Understanding Programs](./3-understanding-programs)

Learn how to understand a larger programs by finding connections between the
details and the big picture. By the end of this chapter you will know how to
read a new program and do a simple code review.

- [ ] 🥚 **Stepping Through**: You can pause a script in a your browser's
      debugger, arrange the debugger, collapse extra panels, and step through a
      script written with Just Enough JS. At each point in execution you can
      make a prediction of the next line before executing, and can check your
      prediction using the _scopes_ panel.
- [ ] 🥚 **Imperative Programming**: You can explain what the Imperative
      Programming paradigm is, and can explain how you know the programs in
      Welcome to JS are Imperative.
- [ ] 🥚 **Logging**: You can tracing specific aspects of a program's execution
      and log them to the console.
- [ ] 🐣 **Naming Variables**: You can analyze how a variable is used in a
      program and give it two names:
  - **Generic**: You can give a generic name to a variable based on how it is
    used in the program.
  - **Specific**: You can give a specific name to a variable based on how it's
    used _and_ the program's domain (the program's specific data and use-case).
- [ ] 🐣 **Describing Programs**: You can read a program and describe it with
      comments using to the methodology from `/describing-programs`:
  1. Zoom Out
  2. Zoom In
  3. Find Connections
  4. Describe Goals
- [ ] 🐥 **Code Review**: Given a working program you can review the code for
      clarity, correctness and style.

**Comment by hadoan149** [comment]
# Week 1

## I Need Help With:

Nothing so far except I'm a bit slow self-studying at the moment. I need to work more on time management!

## What went well?

I think I've got the hang of it. Definitely, the logic of Javascript interests me a lot!

## What went less well?
I just realized I worked better and more efficiently in a group. That explains I'm a bit behind...
Catching up right now!

## Lessons Learned
Check the examples, do the exercises and repeat just don't jump right at the exercises.

## Sunday Prep Work
I'm still at 'what-is-programming'

**Comment by danielhalasz** [comment]
> I just realized I worked better and more efficiently in a group. That explains I'm a bit behind...

you can organize study groups, it can be a good idea to do some learning together

**Comment by danielhalasz** [comment]
> I need to work more on time management!

yes, and please check Discord more often and interact as much as you can

**Comment by hadoan149** [comment]
# Week 3

## I Need Help With:

Nothing so far, however, I did not do all the exercises (I assumed I understood them 😕 ?). I will have more time for the 3rd part on Friday

## What went well?

Understanding why the code was put in a certain order and tracing
## What went less well?
Writing the code myself, so far, I am not confident to write it on my own but I did understand the exercises.

## Lessons Learned
It gets more confusing and I should prepare myself
## Sunday Prep Work
Starting with the new chapter tmr

**Comment by colevandersWands** [comment]
> Writing the code myself, so far, I am not confident to write it on my own but I did understand the exercises.

That's ok, writing your own code is still a 🐔 skill for now.   There will be plenty of time and exercises to learn about writing your own code soon.

**Comment by hadoan149** [comment]
## I Need Help With
For loop but Dnyandeo really helped me out

## What went well?

- The exercises last Sunday! We were able to finish all but two exercises left although we joined online. Our group will meet up today to finish the rest
- debugging!

## What went less well?

- The theory part, I was sick early this week so I could only start reading it now

## Sunday Prep Work

-Finishing up the UX/UI assignment

**Comment by danielhalasz** [comment]
> For loop but Dnyandeo really helped me out

happy to hear that
